docs(session-plugin): docs for session library v4 (#5122)
Kong/docs.konghq.com: app/_hub/kong-inc/session/_index.md

      referenceable: true
      description: |
        The secret that is used in keyed HMAC generation.​
    - name: audience
      minimum_version: "3.2.x"
      required: false
      default: '"default"'
      datatype: string
      description: The session audience, for example "my-application"
    - name: remember
      minimum_version: "3.2.x"
      required: false
      default: false
      datatype: boolean
      description: Enables or disables persistent sessions
    - name: remember_cookie_name
      minimum_version: "3.2.x"
      required: false
      default: '"remember"'
      datatype: string

>       description: Persistent session cookie name

Author: no, if `remember` is set to `true` the default string `remember` will be used as the value of this field (the cookie name).

    - name: remember_rolling_timeout
      minimum_version: "3.2.x"
      required: false
      default: 604800
      datatype: integer

>       description: Persistent session rolling timeout in seconds.

Reviewer: ```suggestion
      description:  The persistent session rolling timeout window, in seconds.
```

    - name: remember_absolute_timeout
      minimum_version: "3.2.x"
      required: false
      default: 2592000
      datatype: integer

>       description: Persistent session absolute timeout in seconds.

Reviewer: ```suggestion
      description: The persistent session absolute timeout limit, in seconds.
```

    - name: request_headers
      minimum_version: "3.2.x"
      required: false
      default: null
      datatype: array of string elements
      description: |

>         List of information to include (as headers) in the request to upstream. 

Reviewer: ```suggestion
        List of information to include, as headers, in the request to the upstream. 
```

        Accepted values are: `id`, `audience`, `subject`, `timeout`, `idling-timeout`, `rolling-timeout`, and
        `absolute-timeout`.

>         For example, { "id", "timeout" } will set both `Session-Id` and `Session-Timeout` in the request headers.

Reviewer: ```suggestion
        For example, `{ "id", "timeout" }` sets both `Session-Id` and `Session-Timeout` in the request headers.
```

    - name: response_headers
      minimum_version: "3.2.x"
      required: false
      default: null
      datatype: array of string elements
      description: |

>         List of information to include (as headers) in the response to downstream. 

Reviewer: ```suggestion
        List of information to include, as headers, in the response to the downstream. 
```

        Accepted values are: `id`, `audience`, `subject`, `timeout`, `idling-timeout`, `rolling-timeout`, and
        `absolute-timeout`.

>         For example: { "id", "timeout" } will inject both `Session-Id` and `Session-Timeout` in the response headers.

Reviewer: ```suggestion
        For example: `{ "id", "timeout" }` injects both `Session-Id` and `Session-Timeout` in the response headers.
```

    - name: cookie_name
      required: false
      default: '`session`'
      datatype: string
      description: The name of the cookie.
    - name: cookie_lifetime
      maximum_version: "3.1.x"
      required: false
      default: 3600
      datatype: number
      description: The duration in seconds that the session will remain open.
    - name: rolling_timeout
      minimum_version: "3.2.x"
      required: false
      default: 3600
      datatype: integer
      description: |

>         The session cookie rolling timeout in seconds.

Reviewer: ```suggestion
        The session cookie rolling timeout, in seconds.
```

        Specifies how long the session can be used until it needs to be renewed.
    - name: absolute_timeout
      minimum_version: "3.2.x"
      required: false
      default: 86400
      datatype: integer
      description: |

>         The session cookie absolute timeout in seconds.

Reviewer: ```suggestion
        The session cookie absolute timeout. in seconds.
```

        Specifies how long the session can be used until it is no longer valid.
    - name: cookie_idletime
      maximum_version: "3.1.x"
      required: false
      datatype: number
      description: |
        The cookie idle time (in seconds); if a cookie is not used for this time
        period, the session becomes invalid. This value is not set by default,
        meaning idle time checks are disabled.
    - name: idling_timeout
      minimum_version: "3.2.x"
      required: false
      default: 900
      datatype: integer

>       description: The session cookie idle time in seconds.

Reviewer: ```suggestion
      description: The session cookie idle time, in seconds.
```

    - name: cookie_renew

>       maximum_version: "3.1.x"

Reviewer: Is this accurate? The changelog entry says the parameter is deprecated, not removed. If it's only deprecated, we should leave it in 3.2.x and add a note to the description saying that it's deprecated, planned to be removed, and to please use X parameter instead.

Author: `Removed` is probably a better fit, I fixed that. The thing is, the field is still accepted without producing errors. if configured, but it is ineffective. `Deprecated` probably suggests it still works though, so that was not good.

      datatype: string
      description: The domain with which the cookie is intended to be exchanged.
    - name: cookie_samesite
      maximum_version: "3.1.x"
      required: false
      default: Strict
      datatype: string
      description: |

>         Determines whether and how a cookie may be sent with cross-site requests. `Strict`:
>         The browser will send cookies only if the request originated from the website that set the cookie.
>         `Lax`: Same-site cookies are withheld on cross-domain subrequests, but will be sent when a user navigates
>         to the URL from an external site, for example, by following a link. `None` or `off`: Disables
>         the same-site attribute so that a cookie may be sent with cross-site requests. `None` requires
>         the Secure attribute (`cookie_secure`) in latest browser versions. For more info, see the
>         [SameSite cookies docs on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite).'

Reviewer: ```suggestion
        Determines whether and how a cookie may be sent with cross-site requests. 
        
        * `Strict`: The browser sends cookies only if the request originated from the website that set the cookie.
        * `Lax`: Same-site cookies are withheld on cross-domain subrequests, but are sent when a user navigates
        to the URL from an external site, for example, by following a link. 
        * `None` or `off`: Disables the same-site attribute so that a cookie may be sent with cross-site requests. 
        `None` requires the Secure attribute (`cookie_secure`) in latest browser versions. For more information, 
        see the [SameSite cookies docs on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite).'
```

        request over the HTTPS protocol. See the
        [Restrict access to cookies docs on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Restrict_access_to_cookies).
    - name: cookie_discard
      maximum_version: "3.1.x"
      required: false
      default: 10
      datatype: number
      description: The duration in seconds after which an old session’s TTL is updated that an old cookie is discarded.
    - name: stale_ttl
      minimum_version: "3.2.x"
      required: false
      default: 10
      datatype: number
      description: |

>       The duration in seconds after which an old cookie is discarded, starting from the moment
>       when the session became outdated and was replaced by a new one.

Reviewer: ```suggestion
      The duration, in seconds, after which an old cookie is discarded, starting from the moment
      when the session becomes outdated and is replaced by a new one.
```

## Changelog

**{{site.base_gateway}} 3.2.x**
* The plugin has been updated to use version 4.0.0 of the `lua-resty-session` library. This introduced several new features, such as the possibility to specify an `audience` for the session.
The following configuration parameters were affected:

Added:
  * `audience`
  * `remember`
  * `remember_cookie_name`
  * `remember_rolling_timeout`
  * `remember_absolute_timeout`
  * `absolute_timeout`
  * `request_headers`
  * `response_headers`

Renamed:
  * `cookie_lifetime` to `rolling_timeout`
  * `cookie_idletime` to `idling_timeout`
  * `cookie_samesite` to `cookie_same_site`
  * `cookie_httponly` to `cookie_http_only`

>   * `cookie_discard` to `stale_ttl`

Reviewer: Question: do we still have backwards compatibility for the renamed parameters, or is this a breaking change, where old configurations will no longer work?

>   * `cookie_discard` to `stale_ttl`

Author: we have backwards compatibility and no plans to deprecate the renamed parameters. All parameters are still accepted with their "old" names. Because of this, from a gateway perspective, I believe there is no need to update the developer portal documentation, apart maybe for removing the `cookie_renew` field, which is now ignored. I am going to test this today, thanks for pointing it out!

>   * `cookie_discard` to `stale_ttl`

Author: @lena-larionova I confirmed that both kong manager and dev portal's sessions still work with or without providing the `cookie_renew` field. It would be a good idea to remove it from all the 3.2.x docs though, like the ones you linked. Should I just go ahead and remove the field from the pages you have linked or is there something to do to version this change correctly?

>   * `cookie_discard` to `stale_ttl`

Reviewer: > Should I just go ahead and remove the field from the pages you have linked or is there something to do to version this change correctly?

Hmm. Good question. I _think_ the answer is simply remove it - from what I can tell, it was only a suggested parameter, and never a required one.

> Added:
>   * `audience`
>   * `remember`
>   * `remember_cookie_name`
>   * `remember_rolling_timeout`
>   * `remember_absolute_timeout`
>   * `absolute_timeout`
>   * `request_headers`
>   * `response_headers`
> 
> Renamed:
>   * `cookie_lifetime` to `rolling_timeout`
>   * `cookie_idletime` to `idling_timeout`
>   * `cookie_samesite` to `cookie_same_site`
>   * `cookie_httponly` to `cookie_http_only`
>   * `cookie_discard` to `stale_ttl`
> 
> Deprecated:
>   * `cookie_renew`

Reviewer: ```suggestion
    Added:
      * `audience`
      * `remember`
      * `remember_cookie_name`
      * `remember_rolling_timeout`
      * `remember_absolute_timeout`
      * `absolute_timeout`
      * `request_headers`
      * `response_headers`
    
    Renamed:
      * `cookie_lifetime` to `rolling_timeout`
      * `cookie_idletime` to `idling_timeout`
      * `cookie_samesite` to `cookie_same_site`
      * `cookie_httponly` to `cookie_http_only`
      * `cookie_discard` to `stale_ttl`
    
    Deprecated:
      * `cookie_renew`
```
